### PR TITLE
compiler: add error for unnecessary use of 'var'

### PR DESCRIPTION
**Pull request by mlugg**
Depends on #18006, which corrects some unnecessary uses of `var` in Aro.

This PR implements #224. The compile error is emitted when a `var` is never used as an lvalue - more detailed analysis is unfortunately not possible under Zig's current analysis model, as whether a `var` is actually mutated may depend on semantic analysis where conditional compilation may incorrect affect the answer.

Despite how limited this analysis is, the change identified around 1000 incorrect uses across the standard library and compiler, which have been manually corrected. It also led to me catching a few bugs in `std`, and one in the compiler - these have been fixed, aside from one in `std` which I wasn't sure of the correct solution for, so a temporary `@compileError` has been put in place to point out the issue. (This shows the usefulness of the error!)

Since the analysis is simply based on whether a variable is ever used as an lvalue, there is a quite convenient way to silence this error when needed: `_ = &x`. This is kind of a dirty secret, and should ideally never be used in user code, but it's useful in the test cases and behavior tests, where we sometimes want to force values to be semantically runtime-known.

The actual implementation of this change is fairly trivial - the bulk of this PR is spent fixing the entire codebase to not trigger this error.

### CODE REVIEW

File: lib/std/math/big/int_test.zig [hunk 1]
```diff
@@ -517,8 +515,8 @@ test "big.int add multi-single" {
 }
 
 test "big.int add multi-multi" {
-    var op1: u128 = 0xefefefef7f7f7f7f;
-    var op2: u128 = 0xfefefefe9f9f9f9f;
+    const op1: u128 = 0xefefefef7f7f7f7f;
+    const op2: u128 = 0xfefefefe9f9f9f9f;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:519]: Often `var` is used to ensure that a value is present at runtime.
So this PR has the unintended side effect of breaking a bunch of test cases, that specifically test runtime behavior.
Changing the `var`s here to `const` for example makes the test case tautological, see also #13923.

**Review by andrewrk — COMMENTED** [review]
Great work! I'm very happy about this change. I've been letting contributions slide that make this mistake knowing that the compiler would solve the problem eventually. Thanks for going through and making the fixes.

File: lib/std/math/big/int_test.zig [hunk 2]
```diff
@@ -831,8 +832,8 @@ test "big.int sub multi-single" {
 }
 
 test "big.int sub multi-multi" {
-    var op1: u128 = 0xefefefefefefefefefefefef;
-    var op2: u128 = 0xabababababababababababab;
+    const op1: u128 = 0xefefefefefefefefefefefef;
+    const op2: u128 = 0xabababababababababababab;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:836]: .

File: lib/std/math/big/int_test.zig [hunk 3]
```diff
@@ -918,8 +919,8 @@ test "big.int mul multi-multi" {
     if (builtin.zig_backend == .stage2_c) return error.SkipZigTest;
     if (builtin.zig_backend == .stage2_x86_64) return error.SkipZigTest;
 
-    var op1: u256 = 0x998888efefefefefefefef;
-    var op2: u256 = 0x333000abababababababab;
+    const op1: u256 = 0x998888efefefefefefefef;
+    const op2: u256 = 0x333000abababababababab;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:923]: .

File: lib/std/math/big/int_test.zig [hunk 4]
```diff
@@ -1040,8 +1041,8 @@ test "big.int mulWrap multi-multi unsigned" {
...
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:1045]: .

File: lib/std/math/big/int_test.zig [hunk 5]
```diff
@@ -1162,8 +1163,8 @@ test "big.int div single-single with rem" {
 }
 
 test "big.int div multi-single no rem" {
-    var op1: u128 = 0xffffeeeeddddcccc;
-    var op2: u128 = 34;
+    const op1: u128 = 0xffffeeeeddddcccc;
+    const op2: u128 = 34;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:1167]: .

File: lib/std/math/big/int_test.zig [hunk 6]
```diff
@@ -1181,8 +1182,8 @@ test "big.int div multi-single no rem" {
 }
 
 test "big.int div multi-single with rem" {
-    var op1: u128 = 0xffffeeeeddddcccf;
-    var op2: u128 = 34;
+    const op1: u128 = 0xffffeeeeddddcccf;
+    const op2: u128 = 34;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:1186]: .

File: lib/std/math/big/int_test.zig [hunk 7]
```diff
@@ -1200,8 +1201,8 @@ test "big.int div multi-single with rem" {
 }
 
 test "big.int div multi>2-single" {
-    var op1: u128 = 0xfefefefefefefefefefefefefefefefe;
-    var op2: u128 = 0xefab8;
+    const op1: u128 = 0xfefefefefefefefefefefefefefefefe;
+    const op2: u128 = 0xefab8;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:1205]: .

File: lib/std/math/big/int_test.zig [hunk 8]
```diff
@@ -2105,7 +2106,7 @@ test "big.int sat shift-left signed simple positive" {
 test "big.int sat shift-left signed multi positive" {
     if (builtin.zig_backend == .stage2_x86_64) return error.SkipZigTest;
 
-    var x: SignedDoubleLimb = 1;
+    const x: SignedDoubleLimb = 1;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:2109]: .

File: lib/std/math/big/int_test.zig [hunk 9]
```diff
@@ -2118,7 +2119,7 @@ test "big.int sat shift-left signed multi positive" {
 test "big.int sat shift-left signed multi negative" {
     if (builtin.zig_backend == .stage2_x86_64) return error.SkipZigTest;
 
-    var x: SignedDoubleLimb = -1;
+    const x: SignedDoubleLimb = -1;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:2122]: .

File: lib/std/math/big/int_test.zig [hunk 10]
```diff
@@ -2129,7 +2130,7 @@ test "big.int sat shift-left signed multi negative" {
 }
 
 test "big.int bitNotWrap unsigned simple" {
-    var x: u10 = 123;
+    const x: u10 = 123;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:2133]: .

File: lib/std/math/big/int_test.zig [hunk 11]
```diff
@@ -2148,7 +2149,7 @@ test "big.int bitNotWrap unsigned multi" {
 }
 
 test "big.int bitNotWrap signed simple" {
-    var x: i11 = -456;
+    const x: i11 = -456;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:2152]: .

File: lib/std/math/big/int_test.zig [hunk 12]
```diff
@@ -2304,8 +2305,8 @@ test "big.int bitwise xor simple" {
 }
 
 test "big.int bitwise xor multi-limb" {
-    var x: DoubleLimb = maxInt(Limb) + 1;
-    var y: DoubleLimb = maxInt(Limb);
+    const x: DoubleLimb = maxInt(Limb) + 1;
+    const y: DoubleLimb = maxInt(Limb);
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:2309]: .

File: lib/std/math/big/int_test.zig [hunk 13]
```diff
@@ -2799,7 +2800,7 @@ test "big int conversion read twos complement with padding" {
     // (2) should correctly order bytes based on the provided endianness
     // (3) should sign-extend any bits from bit_count to 8 * abi_size
 
-    var bit_count: usize = 12 * 8 + 1;
+    const bit_count: usize = 12 * 8 + 1;
```
Comment (IntegratedQuantum) [inline lib/std/math/big/int_test.zig:2803]: Also this doesn't seem right. A few lines down in line 2815 this variable seems to be modified:
```zig
bit_count = 12 * 8 + 2;
```

**Review by IntegratedQuantum — COMMENTED** [review]
More tautological cases you missed in math/big_int_test.zig

**Comment by InKryption** [comment]
A small issue I've just encountered with this change is that unlike with `_ = x;`, `_ = &x;` doesn't become an error after it becomes unnecessary. If `_ = &x;` is to become the official way to allow unused mutability, then it should probably be recognized as unnecessary as well.

**Comment by Atomk** [comment]
In my opinion `_ = &x;` looks a bit too much like a hack to be the "official" way to do it, maybe something like a `@forceRuntimeKnown` that can only be used in tests would help to clarify the intent.

**Comment by andrewrk** [comment]
The official way is to fix your code. Don't work around the error.

**Comment by mlugg** [comment]
The workaround has the following uses:
* working around compiler bugs where making something `const` triggers the bug
* writing tests for the compiler itself (behavior tests or test cases)
* writing std tests where you absolutely need to force a value to be runtime-known; this came up in bigint tests and a small handful of other places

This means that outside of the compiler codebase, there is pretty much no use in it.

Also note: currently, `translate-c` emits `_ = &x` discards all over the place to avoid triggering the error, but I hope that eventually we can instead internally give the translate-c output to `zig fmt --autofix`, so things just become `const` (which is the simple, correct, readable solution).

**Comment by linusg** [comment]
This has a limitation that I'd assume to be quite common as it applies to built-ins like `@memset()`:

```zig
fn foo(ptr: *usize) void {
    ptr.* = 42;
}

fn main() void {
    const s: struct { i: usize = 0 } = .{};
    var i = &s.i; // error: local variable is never mutated
    foo(i);
}
```

**Comment by andrewrk** [comment]
The compiler is correct; the code is wrong. It should be `const` not `var`.

**Comment by linusg** [comment]
That doesn't build either (`0.12.0-dev.1676+40b8c993f`):

```
file.zig:8:9: error: expected type '*usize', found '*const usize'
    foo(i);
```

**Comment by Snektron** [comment]
You mean to write
```zig
fn foo(ptr: *usize) void {
    ptr.* = 42;
}

fn main() void {
    var s: struct { i: usize = 0 } = .{};
    const i = &s.i;
    foo(i);
}
```

**Pull request merged**